File: solutions/Hard/Binary_Tree_Maximum_Path_Sum_0124.py
from typing import Optional

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        """
        Function: maxPathSum
        Description: Finds the maximum path sum of any non-empty path in a binary tree.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.

        Returns:
        - int: The maximum path sum.
        """
        self.max_sum = float('-inf')

        def dfs(node):
            if not node:
                return 0

            # Compute the max path sum for left and right subtrees
            left_max = max(dfs(node.left), 0)  # Ignore negative paths
            right_max = max(dfs(node.right), 0)

            # Compute the max path sum passing through the current node
            current_path_sum = node.val + left_max + right_max

            # Update global max path sum
            self.max_sum = max(self.max_sum, current_path_sum)

            # Return the max sum of path ending at the current node
            return node.val + max(left_max, right_max)

        dfs(root)

        return self.max_sum


# Helper functions for testing
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in lst]

    for i, node in enumerate(nodes):
        if node:
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst):
                node.left = nodes[left_idx]
            if right_idx < len(lst):
                node.right = nodes[right_idx]

    return nodes[0]  # Root node


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Simple tree
    root = list_to_binary_tree([1, 2, 3])
    print(solution.maxPathSum(root))
    # Expected output: 6 (2 -> 1 -> 3)

    # Test case 2: Tree with negative values
    root = list_to_binary_tree([-10, 9, 20, None, None, 15, 7])
    print(solution.maxPathSum(root))
    # Expected output: 42 (15 -> 20 -> 7)

    # Test case 3: Single node tree
    root = list_to_binary_tree([5])
    print(solution.maxPathSum(root))
    # Expected output: 5

    # Test case 4: Tree with only negative numbers
    root = list_to_binary_tree([-3])
    print(solution.maxPathSum(root))
    # Expected output: -3

    # Test case 5: Complex tree with mixed values
    root = list_to_binary_tree(
        [10, 2, 10, 20, 1, None, -25, None, None, None, None, None, None, 3, 4])
    print(solution.maxPathSum(root))
    # Expected output: 42 (20 -> 2 -> 10 -> 10)

    # Test case 6: Large tree
    root = list_to_binary_tree(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15])
    print(solution.maxPathSum(root))
    # Expected output: 60 (15 -> 14 -> 13 -> 12 -> 11 -> 10 -> 9 -> 8 -> 7 -> 6)

    # Test case 7: Large tree with negative values
    root = list_to_binary_tree(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, -15])
    print(solution.maxPathSum(root))
    # Expected output: 56 (14 -> 13 -> 12 -> 11 -> 10 -> 9 -> 8 -> 7 -> 6)

    # Test case 8: Large tree with negative values
    root = list_to_binary_tree(
        [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, -13, -14, -15])
    print(solution.maxPathSum(root))
    # Expected output: 56 (11 -> 10 -> 9 -> 8 -> 7 -> 6)

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(N), where N is the number of nodes in the binary tree. We visit each node once.
# Space Complexity: O(H), where H is the height of the binary tree. This space is used for the recursive call stack.
# In the worst case, the height of the binary tree is O(N), leading to a space complexity of O(N). In the average case, the space complexity is O(log(N)).
# In the worst case, the space complexity is O(N) for a skewed binary tree.
# In the average case, the space complexity is O(log(N)) for a balanced binary tree.

File: solutions/Hard/Binary_Tree_Maximum_Path_Sum_0124.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val);
 *     this.left = (left===undefined ? null : left);
 *     this.right = (right===undefined ? null : right);
 * }
 */

/**
 * Function: maxPathSum
 * Description: Given the root of a binary tree, this function returns the maximum path sum
 * for any path in the tree. A path consists of nodes connected by edges, and each node 
 * can only appear once in a path.
 *
 * @param {TreeNode} root - The root of the binary tree.
 * @returns {number} - The maximum path sum of any non-empty path.
 */
const maxPathSum = function(root) {
    let maxSum = -Infinity;

    /**
     * Helper function to compute maximum path sum for each subtree.
     * @param {TreeNode} node - The current node.
     * @returns {number} - Maximum path sum including this node.
     */
    const maxGain = (node) => {
        if (!node) return 0;

        // Compute max sum on the left and right; discard negatives (max with 0)
        let leftGain = Math.max(maxGain(node.left), 0);
        let rightGain = Math.max(maxGain(node.right), 0);

        // Compute max path sum where current node is the highest node (split path)
        let pathSum = node.val + leftGain + rightGain;

        // Update max sum found
        maxSum = Math.max(maxSum, pathSum);

        // Return max gain from this node (choosing either left or right)
        return node.val + Math.max(leftGain, rightGain);
    };

    maxGain(root);
    
    return maxSum;
};

// Example Test Cases
if (require.main === module) {
    const TreeNode = function(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    };

    console.log("Test 1: Basic case with root = [1,2,3]");
    let root1 = new TreeNode(1, new TreeNode(2), new TreeNode(3));
    console.log(maxPathSum(root1));
    // Expected output: 6 (Path: 2 -> 1 -> 3)

    console.log("Test 2: More complex tree");
    let root2 = new TreeNode(-10, 
        new TreeNode(9), 
        new TreeNode(20, new TreeNode(15), new TreeNode(7))
    );
    console.log(maxPathSum(root2));
    // Expected output: 42 (Path: 15 -> 20 -> 7)

    console.log("Test 3: Single node tree");
    let root3 = new TreeNode(5);
    console.log(maxPathSum(root3));
    // Expected output: 5

    console.log("Test 4: Tree with all negative values");
    let root4 = new TreeNode(-3, new TreeNode(-2), new TreeNode(-1));
    console.log(maxPathSum(root4));
    // Expected output: -1 (Max single node)

    console.log("Test 5: Tree with zero values");
    let root5 = new TreeNode(0, new TreeNode(-2), new TreeNode(3));
    console.log(maxPathSum(root5));
    // Expected output: 3 (Choosing the right child only)

    console.log("Test 6: Unbalanced tree");
    let root6 = new TreeNode(1, new TreeNode(2, new TreeNode(3, new TreeNode(4))), null);
    console.log(maxPathSum(root6));
    // Expected output: 10 (Path: 4 -> 3 -> 2 -> 1)
}

/*
Time Complexity: O(n) - Visits each node once.
Space Complexity: O(h) - Recursion depth equals the tree height (O(log n) for balanced trees, O(n) for skewed trees).
*/

export default maxPathSum;
